### PR TITLE
feat: support audit log filter http api

### PR DESCRIPTION
**Pull request by zhongwencool**
Fixes https://emqx.atlassian.net/browse/EMQX-11173 and https://emqx.atlassian.net/browse/EMQX-10928
image

### <samp>🤖 Generated by Copilot at 63b0c25</samp>

This pull request adds the `emqx_audit` application for audit logging in EMQX, and refactors the existing audit logging mechanism with a new macro `LOG_AUDIT_EVENT`. It also adds the `permissions` key to various GitHub workflow files to limit the access of the GitHub actions runner.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Comment by JimMoen** [comment]
reopen to trigger CI

File: apps/emqx_management/src/emqx_mgmt_api.erl [hunk 1]
```diff
@@ -341,11 +341,11 @@ do_select(
         try
             case maps:get(continuation, QueryState, undefined) of
                 undefined ->
-                    ets:select(Tab, Ms, Limit);
+                    ets:select_reverse(Tab, Ms, Limit);
```
Comment (zhongwencool) [inline apps/emqx_management/src/emqx_mgmt_api.erl:344]: https://www.erlang.org/doc/man/ets#select_reverse-3
> Works like [select/3](https://www.erlang.org/doc/man/ets#select-3), but for table type > ordered_set traversing is done starting at the last object in Erlang term order and moves to the first. For all other table types, the return value is identical to that of select/3.

> Notice that this is not equivalent to reversing the result list of a select/3 call, as the result list is not only reversed, but also contains the last Limit matching objects in the table, not the first.

File: apps/emqx_audit/include/emqx_audit.hrl [hunk 2]
```diff
@@ -0,0 +1,39 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (zmstone) [inline apps/emqx_audit/include/emqx_audit.hrl:8]: The app is under BSL this file should not be APL?

Comment (zmstone) [reply apps/emqx_audit/include/emqx_audit.hrl:8]: Same applies for other files.

**Review by zmstone — COMMENTED** [review]
Just had a very quick look . Not done yet.

File: rel/i18n/emqx_audit_api.hocon [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+filter_node.desc:
+"Filter logs based on the node name to which the logs are created."
+
+filter_from.desc:
+""""Filter logs based on source type, valid values include:
+`dashboard`: Dashboard request logs, requiring the use of a jwt_token.
+`rest_api`: API KEY request logs.
+`cli`: The emqx command line logs.
+`erlang_console`: The emqx remote_console run function logs.
+`event`: Logs related to events such as emqx_start, emqx_stop, audit_enabled, and audit_disabled."""
```
Comment (HJianBo) [inline rel/i18n/emqx_audit_api.hocon:20]: nit:
```suggestion
`event`: Logs related to events such as emqx_start, emqx_gracefully_stop, audit_enabled, and audit_disabled."""
```

File: apps/emqx_audit/src/emqx_audit_api.erl [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+api_spec() ->
+    emqx_dashboard_swagger:spec(?MODULE, #{check_schema => true}).
+
+paths() ->
+    ["/audit"].
+
+schema("/audit") ->
+    #{
+        'operationId' => audit,
```
Comment (HJianBo) [inline apps/emqx_audit/src/emqx_audit_api.erl:49]: Q: Do we need to support enabling/disabling audit via HTTP API, and get the enabled status via HTTP API? 🤔

For the current implementation, we only support enable/disable this feature via a config file (Or CLI to update config)

File: apps/emqx_dashboard/src/emqx_dashboard_audit.erl [hunk 5]
```diff
@@ -18,29 +18,98 @@
...
-level(get, _Code, _) -> debug;
-level(_, Code, _) when Code >= 200 andalso Code < 300 -> info;
-level(_, Code, _) when Code >= 300 andalso Code < 400 -> warning;
-level(_, Code, _) when Code >= 400 andalso Code < 500 -> error;
-level(_, _, _) -> critical.
+-export([log/2]).
+
+%% filter high frequency events
+-define(HIGH_FREQUENCY_REQUESTS, [
+    <<"/clients/:clientid/publish">>,
```
Comment (HJianBo) [inline apps/emqx_dashboard/src/emqx_dashboard_audit.erl:25]: nit: 🤔 As simple testing, the operation_id of publish API `POST /publish` is
```
{..., "operation_id":"/publish", ...}
```

Comment (zhongwencool) [reply apps/emqx_dashboard/src/emqx_dashboard_audit.erl:25]: 😢 sorry,     bad gpt completion

File: mix.exs [hunk 6]
```diff
@@ -214,7 +214,8 @@ defmodule EMQXUmbrella.MixProject do
       :emqx_bridge_azure_event_hub,
       :emqx_gcp_device,
       :emqx_dashboard_rbac,
-      :emqx_dashboard_sso
+      :emqx_dashboard_sso,
+      :emqx_audit
```
[the remark this answers is not in the document]
Comment (zhongwencool) [reply mix.exs:218]: Good catch!

File: apps/emqx_audit/src/emqx_audit_api.erl (hunk 4 above)
Comment (zhongwencool) [reply apps/emqx_audit/src/emqx_audit_api.erl:49]: It's part on `log` configuration. we already have `/config/log` api to update.

**Review by HJianBo — APPROVED** [review]
LGTM

**Pull request merged**